File: solutions/javascript/perfect-numbers/1/perfect-numbers.js
export const classify = (candidate) => {
  if (candidate <= 0)
    throw new Error("Classification is only possible for natural numbers.");

  const primes = new Set();
  let high = candidate;
  for (let low = 1; low <= high; low++) {
    if (candidate % low == 0) {
      high = candidate / low;
      primes.add(low);
      primes.add(high);
    }
  }
  primes.delete(candidate);

  let aliquotSum = 0;
  primes.forEach((prime) => {
    aliquotSum += prime;
  });

  if (aliquotSum == candidate) return "perfect";
  else if (aliquotSum > candidate) return "abundant";
  else return "deficient";
};

File: solutions/python/perfect-numbers/1/perfect_numbers.py
import math
from typing import Literal

def factors_excluding_self(number) -> set[int]:
    factors = set()

    if number > 1:
        factors.add(1)
        
    i = 2
    ceil = number // 2
    while i <= ceil:
        if number % i == 0:
            factors.add(i)
            factors.add(number // i)
            ceil = number // i
        i += 1
    return factors

def classify(number: int) -> Literal["perfect", "abundant", "deficient"]:
    """ A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """

    if number < 1:
        raise ValueError("Classification is only possible for positive integers.")

    aliquot_sum = sum(factors_excluding_self(number))

    if aliquot_sum == number:
        return "perfect"
    elif aliquot_sum > number:
        return "abundant"
    else:
        return "deficient"
